dvc/remote/base.py

```diff
@@ -352,7 +363,49 @@ def changed(self, path_info, checksum_info):
         return False
 
     def link(self, from_info, to_info):
-        self.copy(from_info, to_info)
+        self._link(from_info, to_info, self.cache_types)
+
+    def _link(self, from_info, to_info, link_types):
+        assert self.isfile(from_info)
+
+        self.makedirs(to_info.parent)
+
+        self._try_links(from_info, to_info, link_types)
+
+    @slow_link_guard
+    def _try_links(self, from_info, to_info, link_types):
+        i = len(link_types)
+        while i > 0:
+            link_method = getattr(self, link_types[0])
+            try:
+                self._do_link(from_info, to_info, link_method)
+                return
+
+            except DvcException as exc:
+                msg = "Cache type '{}' is not supported: {}"
+                logger.debug(msg.format(link_types[0], str(exc)))
+                del link_types[0]
+                i -= 1
+
+        raise DvcException("no possible cache types left to try out.")
+
+    def _do_link(self, from_info, to_info, link_method):
+        # XXX: We are testing if file exists rather than if file is a link
+        if self.exists(to_info):
+            raise DvcException("Link '{}' already exists!".format(to_info))
+        else:
```

Comment:
This `else` is not needed.

Reply:
Oops, crossed comments:

https://github.com/iterative/dvc/pull/2426#discussion_r319731653

> I was pointing out that we weren't testing for an actual "link" because the error message says something like: `Link 'foo.txt' already exists!` I guess it is `File` instead of `Link` :sweat_smile:

Reply:
wait, my bad, I got confused; ignore the above one, @efiop